### PR TITLE
Don't fail the sync if there are failed operations in the bulk api request

### PR DESCRIPTION
**Pull request by wangch079**
## Closes https://github.com/elastic/enterprise-search-team/issues/5808

This PR makes 2 changes:
1. When there's field `_id` missing in a document, skip the document and log an error message
2. When one or more operations fail in a bulk request, don't fail the whole sync job, log the error message instead.

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)

### CODE REVIEW

File: connectors/es/sink.py [hunk 1]
```diff
@@ -164,7 +164,6 @@ async def _bulk_api_call():
                 for op, data in item.items():
                     if "error" in data:
                         self._logger.error(f"operation {op} failed, {data['error']}")
-                        raise Exception(data["error"]["reason"])
```
Comment (timgrein) [inline connectors/es/sink.py:167]: Should we adapt the log line above to something like this: `self._logger.error(f"Operation: {op} failed. Reason: {data['error']['reason']}. Full error: {data['error']}")`. Could be easier to spot the error reason faster instead of looking through the whole error object

**Review by artem-shelkovnikov — COMMENTED** [review]
I'm not sure it's a way to do TBH - documents can fail due to a variety of reasons, such as:

1. Mapping is incompatible with document schema
2. Document too big
3. 429

429 should be retried, others not. Just skipping errors breaks our resilience, so we need to parse the reason and retry some, while skipping/logging others.

File: connectors/es/sink.py [hunk 2]
```diff
@@ -164,7 +164,6 @@ async def _bulk_api_call():
                 for op, data in item.items():
                     if "error" in data:
                         self._logger.error(f"operation {op} failed, {data['error']}")
```
Comment (seanstory) [inline connectors/es/sink.py:166]: I think we should count how often this happens, and at least log out the total at the end of the sync. The customer needs to know of 99% of their documents fail here, especially if we start calling such sync jobs "complete" instead of "error"

**Comment by wangch079** [comment]
>I think we should count how often this happens, and at least log out the total at the end of the sync. The customer needs to know of 99% of their documents fail here, especially if we start calling such sync jobs "complete" instead of "error"

Hi @seanstory , this reminds me of the `monitor` object in the Workplace Search, which tracks the error rate and raises if it reaches a certain threshold.

>429 should be retried, others not. Just skipping errors breaks our resilience, so we need to parse the reason and retry some, while skipping/logging others.

The Bulk API request will be retried for [these response codes](https://github.com/elastic/connectors/blob/f4d65e252bae32d228f0e776e517977f5d2ba5ab/connectors/es/client.py#L204), we yet have the retry logic for a single operation in the Bulk API request.

From the [Response body schema](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html#bulk-api-response-body), there is `status` field in each item. We can follow the same [retriable response codes](https://github.com/elastic/connectors/blob/f4d65e252bae32d228f0e776e517977f5d2ba5ab/connectors/es/client.py#L204).


So in conclusion, we need to
1) Add retry logic to operations in Bulk API requests. We need to add a counter to each operation, and only raise if it reaches a threshold or the error is not retriable
2) Add a monitor object (to both extractor and sink), and when an error happens, we call `monitor.note_error`, and let `monitor` decide when to raise an error.

This sounds beyond the scope of this issue, but I just want to make sure we agree on the approach.

**Comment by seanstory** [comment]
> This sounds beyond the scope of this issue, but I just want to make sure we agree on the approach.

works for me.

**Comment by wangch079** [comment]
Hi @artem-shelkovnikov , the PR introduces 2 changes (added in the PR description), and only the first one is related to issue https://github.com/elastic/enterprise-search-team/issues/5808. Let me know if you still have concerns for the second change, and I can exclude it from this PR

**Review by artem-shelkovnikov — APPROVED** [review]

**Comment by artem-shelkovnikov** [comment]
> Hi @artem-shelkovnikov , the PR introduces 2 changes (added in the PR description), and only the first one is related to issue https://github.com/elastic/enterprise-search-team/issues/5808. Let me know if you still have concerns for the second change, and I can exclude it from this PR

Just to check, I see that `raise` was removed, but where does it get logged? What happens here, do we retry the batch if we have errors? Do we log all errors?

**Comment by wangch079** [comment]
> Just to check, I see that `raise` was removed, but where does it get logged?

It's logged [here](https://github.com/elastic/connectors/pull/2271/files#diff-68c31d9379d91378705cf6d47c447185e8c6ad952876c7a5fb979e3c89646148R200)

>What happens here, do we retry the batch if we have errors?

No, the current retry is on Bulk API request, i.e. only if the Bulk API fails entirely. Retry on a single operation is not in scope of this PR, see discussion here: https://github.com/elastic/connectors/pull/2271#issuecomment-2011320554
